Sources/MapboxDirections/Directions.swift

```diff
@@ -152,54 +113,71 @@ open class Directions: NSObject {
      - parameter completionHandler: The closure (block) to call with the resulting routes. This closure is executed on the application’s main thread.
      - returns: The data task used to perform the HTTP request. If, while waiting for the completion handler to execute, you no longer want the resulting routes, cancel this task.
      */
-    @discardableResult open func calculate(_ options: RouteOptions, completionHandler: @escaping RouteCompletionHandler) -> URLSessionDataTask {
-        let fetchStart = Date()
+    @discardableResult open func calculate(_ options: RouteOptions, completionHandler: @escaping RouteCompletionHandler /* FIXME: VARIENT TYPE */) -> URLSessionDataTask {
+        options.fetchStartDate = Date()
         let request = urlRequest(forCalculating: options)
         let requestTask = URLSession.shared.dataTask(with: request) { (possibleData, possibleResponse, possibleError) in
-            let responseEndDate = Date()
-            guard let response = possibleResponse, ["application/json", "text/html"].contains(response.mimeType) else {
-                completionHandler(nil, nil, .invalidResponse)
+            guard let response = possibleResponse, ["application/json", "text/html"].contains(response.mimeType), let httpResponse = response as? HTTPURLResponse else {
+                let response = RouteResponse(httpResponse: possibleResponse as? HTTPURLResponse, options: .route(options), credentials: self.credentials)
+                completionHandler(response, .invalidResponse(possibleResponse))
                 return
             }
             
             guard let data = possibleData else {
-                completionHandler(nil, nil, .noData)
+                let response = RouteResponse(httpResponse: httpResponse, options: .route(options), credentials: self.credentials)
+                completionHandler(response, .noData)
                 return
             }
             
             if let error = possibleError {
-                completionHandler(nil, nil, .unknown(response: possibleResponse, underlying: error, code: nil, message: nil))
+                let response = RouteResponse(httpResponse: httpResponse, options: .route(options), credentials: self.credentials)
+                let unknownError = DirectionsError.unknown(response: possibleResponse, underlying: error, code: nil, message: nil)
+                completionHandler(response, unknownError)
                 return
             }
```

Comment:
Current error handling and this one has big problem: when there is no internet, or request was cancelled, it always returns `invalidResponse` error instead of unknown or other one. This is caused by not checking `possibleError` first. It should be inspected as first one (as it contains error when request is canceled or failed for other reasons - see Foundation.URLError for reference). Only after assuring it is nil, we should inspect `possibleResponse` and `possibleData`.

Reply:
Great suggestion, @SebastianOsinski! Will resolve.